### PR TITLE
Support fisheye projection

### PR DESCRIPTION
**Pull request by levinli303**
Basically adding the same things to every shader, is there a better approach?  Values that depend on FOVs also needs to be worked on.

### CODE REVIEW

**Comment by 375gnu** [comment]
i think this should look as following:
1. fisheye function is in a separate file, e.g. `shaders/fisheye.h`
2. after `#version ...` we add `#define FISHEYE` if fish eye projection is required
3. then we add `#ifdef FISHEYE\n#include <fisheye.h>\n#endif`
4. in the main body `if (fisheye) { ... } else { ... }` is replaced with `#ifdef FISHEYE\nfisheye(inPos)\n#else\n...\n#endif`
5. in `shaderproperties.cpp` in `bool operator<(const ShaderProperties& p0, const ShaderProperties& p1)` add `if (p0.fisheye != p1.fisheye) return int(p0.fisheye) < int(p1.fisheye);`

**Comment by levinli303** [comment]
@375gnu  can we use include in GLSL?

**Comment by 375gnu** [comment]
Forgot to say that glsl doesn't support `#include` so the usual way is to check if a file contains it and replace with actual content.

**Comment by levinli303** [comment]
Finished a dumb version where fisheye is set ShaderManager. for shaderproperties  we could have a property like that, but for shaders from files, I haven't found a viable way

File: src/celestia/celestiacore.cpp [hunk 1]
```diff
@@ -79,8 +79,8 @@ static const float RotationBraking = 10.0f;
 static const float RotationDecay = 2.0f;
 static const double MaximumTimeRate = 1.0e15;
 static const double MinimumTimeRate = 1.0e-15;
-static const float stdFOV = degToRad(45.0f);
-static const float MaximumFOV = degToRad(120.0f);
+static const float stdFOV = degToRad(60.0f);
```
Comment (375gnu) [inline src/celestia/celestiacore.cpp:82]: shouldn't this be controlled by fish eye mode?

File: src/celestia/celestiacore.cpp [hunk 2]
```diff
@@ -2284,8 +2284,13 @@ void CelestiaCore::setFOVFromZoom()
     for (const auto v : views)
         if (v->type == View::ViewWindow)
         {
-            double fov = 2 * atan(height * v->height / (screenDpi / 25.4) / 2. / distanceToScreen) / v->zoom;
-            v->observer->setFOV((float) fov);
+            if (renderer->getProjectionMode() == Renderer::FisheyeMode)
+                v->observer->setFOV((float) MaximumFOV);
```
Comment (375gnu) [inline src/celestia/celestiacore.cpp:2288]: MaximumFOV is already float

File: src/celengine/shadermanager.cpp [hunk 3]
```diff
@@ -68,10 +68,33 @@ static const char* VertexHeader = R"glsl(
 uniform mat4 ModelViewMatrix;
 uniform mat4 ProjectionMatrix;
 uniform mat4 MVPMatrix;
+uniform bool fisheye;
```
Comment (375gnu) [inline src/celengine/shadermanager.cpp:71]: this uniform is not used anywhere

File: src/celengine/shadermanager.cpp [hunk 4]
```diff
@@ -3383,6 +3438,11 @@ CelestiaGLProgram::intParam(const string& paramName)
     return IntegerShaderParameter(program->getID(), paramName.c_str());
 }
 
+BooleanShaderParameter
```
Comment (375gnu) [inline src/celengine/shadermanager.cpp:3441]: BooleanShaderParameter is not used anywhere

**Review by 375gnu — CHANGES_REQUESTED** [review]

**Comment by levinli303** [comment]
should have fixed all the issues. also disabled the star glare rendering on fisheye for now

File: src/celengine/observer.cpp [hunk 5]
```diff
@@ -1352,6 +1352,21 @@ Vector3f Observer::getPickRay(float x, float y) const
     return pickDirection.normalized();
 }
 
+Vector3f Observer::getPickRayFisheye(float x, float y) const
+{
+    float r = sqrt(x * x + y * y);
```
Comment (375gnu) [inline src/celengine/observer.cpp:1357]: you can replace with `hypot` here

**Review by 375gnu — CHANGES_REQUESTED** [review]

File: src/celengine/observer.cpp (hunk 5 above)
Comment (levinli303) [reply src/celengine/observer.cpp]: fixed

**Comment by levinli303** [comment]
let's merge this

**Review by 375gnu — APPROVED** [review]
let's merge

**Pull request merged**